File: html_to_markdown/options.py
"""Configuration options for HTML to Markdown conversion.

This module provides dataclass-based configuration for the v2 API.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, Literal, Protocol

if TYPE_CHECKING:
    from collections.abc import Callable

    from bs4 import Tag


class ConverterFunction(Protocol):
    """Protocol for custom converter functions.

    Converter functions receive keyword-only arguments including the HTML tag,
    processed text content, and any conversion options needed.

    Example:
        >>> def custom_link_converter(*, tag: Tag, text: str, autolinks: bool, **kwargs: Any) -> str:
        ...     href = tag.get("href", "")
        ...     return f"[{text}]({href})"
    """

    def __call__(self, *, tag: Tag, text: str, **kwargs: Any) -> str:
        """Convert an HTML element to Markdown.

        Args:
            tag: BeautifulSoup Tag object representing the HTML element
            text: Processed text content of the element's children
            **kwargs: Additional conversion options (varies by converter)

        Returns:
            Markdown string representation of the element
        """
        ...


@dataclass
class ConversionOptions:
    """Main conversion configuration.

    This class groups all conversion-related options together, replacing
    the large number of keyword arguments in the v1 API.

    Example:
        >>> options = ConversionOptions(
        ...     heading_style="atx",
        ...     list_indent_width=2,
        ...     escape_asterisks=True,
        ... )
        >>> from html_to_markdown import convert
        >>> markdown = convert("<h1>Title</h1>", options)
    """

    # Heading options
    heading_style: Literal["underlined", "atx", "atx_closed"] = "underlined"
    """Style for headings: 'underlined' (===), 'atx' (#), or 'atx_closed' (# #)."""

    # List options
    list_indent_type: Literal["spaces", "tabs"] = "spaces"
    """Type of indentation for lists."""

    list_indent_width: int = 4
    """Number of spaces for list indentation (ignored if list_indent_type='tabs')."""

    bullets: str = "*+-"
    """Characters to use for unordered list bullets, cycling through levels."""

    # Text formatting
    strong_em_symbol: Literal["*", "_"] = "*"
    """Symbol for strong/emphasis formatting."""

    escape_asterisks: bool = True
    """Escape asterisk characters in text to prevent accidental formatting."""

    escape_underscores: bool = True
    """Escape underscore characters in text to prevent accidental formatting."""

    escape_misc: bool = True
    """Escape miscellaneous Markdown characters."""

    # Code blocks
    code_language: str = ""
    """Default language for code blocks."""

    code_language_callback: Callable[[Tag], str] | None = None
    """Callback to determine code language from element."""

    # Links
    autolinks: bool = True
    """Convert bare URLs to automatic links."""

    default_title: bool = False
    """Add a default title if none exists."""

    # Images
    keep_inline_images_in: set[str] | None = None
    """Parent tag names where images should remain inline."""

    # Tables
    br_in_tables: bool = False
    """Use <br> tags for line breaks in table cells instead of spaces."""

    # hOCR table extraction
    hocr_extract_tables: bool = True
    """Enable table extraction from hOCR (HTML-based OCR) documents."""

    hocr_table_column_threshold: int = 50
    """Pixel threshold for detecting column boundaries in hOCR tables."""

    hocr_table_row_threshold_ratio: float = 0.5
    """Row height ratio threshold for detecting row boundaries in hOCR tables."""

    # Highlighting
    highlight_style: Literal["double-equal", "html", "bold"] = "double-equal"
    """Style for highlighting <mark> elements."""

    # Metadata
    extract_metadata: bool = True
    """Extract metadata from HTML head and include as comment."""

    # Whitespace
    whitespace_mode: Literal["normalized", "strict"] = "normalized"
    """How to handle whitespace: 'normalized' or 'strict'."""

    strip_newlines: bool = False
    """Remove newlines from HTML before processing."""

    # Wrapping
    wrap: bool = False
    """Enable text wrapping."""

    wrap_width: int = 80
    """Column width for text wrapping."""

    # Element filtering
    convert: set[str] | None = None
    """HTML tags to convert to Markdown (None = all supported tags)."""

    strip: set[str] | None = None
    """HTML tags to strip from output."""

    convert_as_inline: bool = False
    """Treat block elements as inline during conversion."""

    # Subscript/superscript
    sub_symbol: str = ""
    """Symbol for subscript text."""

    sup_symbol: str = ""
    """Symbol for superscript text."""

    # Newlines
    newline_style: Literal["spaces", "backslash"] = "spaces"
    """Style for newlines: 'spaces' (two spaces) or 'backslash' (\\)."""

    # Custom converters
    custom_converters: dict[str, Callable[..., str]] | None = None
    """Custom converter functions for specific HTML elements."""


@dataclass
class PreprocessingOptions:
    """HTML preprocessing configuration.

    Controls how HTML is cleaned and preprocessed before conversion.

    Example:
        >>> options = PreprocessingOptions(
        ...     enabled=True,
        ...     preset="aggressive",
        ...     remove_navigation=True,
        ... )
    """

    enabled: bool = False
    """Whether to enable HTML preprocessing."""

    preset: Literal["minimal", "standard", "aggressive"] = "standard"
    """Preprocessing aggressiveness level."""

    remove_navigation: bool = True
    """Remove navigation elements during preprocessing."""

    remove_forms: bool = True
    """Remove form elements during preprocessing."""

    excluded_navigation_classes: set[str] | None = None
    """Navigation class fragments to keep even when removing navigation."""

    extra_navigation_classes: set[str] | None = None
    """Additional navigation class fragments to strip beyond defaults."""


@dataclass
class ParsingOptions:
    """HTML parsing configuration.

    Example:
        >>> options = ParsingOptions(
        ...     encoding="utf-8",
        ...     detect_encoding=True,
        ... )
    """

    encoding: str = "utf-8"
    """Character encoding for decoding bytes input."""

    detect_encoding: bool = False
    """Attempt to detect encoding from HTML (not yet implemented)."""

    parser: str | None = None
    """HTML parser to use: 'html.parser', 'lxml', or 'html5lib' (None = auto)."""

File: crates/html-to-markdown-wasm/src/options.rs
use crate::enums::{
    WasmCodeBlockStyle, WasmHeadingStyle, WasmHighlightStyle, WasmLinkStyle, WasmListIndentType, WasmNewlineStyle,
    WasmOutputFormat, WasmPreprocessingPreset, WasmWhitespaceMode,
};
use html_to_markdown_rs::{ConversionOptionsUpdate, PreprocessingOptionsUpdate};
use serde::{Deserialize, Serialize};

/// HTML preprocessing options
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct WasmPreprocessingOptions {
    /// Enable preprocessing
    #[serde(default)]
    pub enabled: bool,
    /// Preprocessing preset
    #[serde(default)]
    pub preset: Option<WasmPreprocessingPreset>,
    /// Remove navigation elements
    #[serde(default = "default_true")]
    pub remove_navigation: bool,
    /// Remove form elements
    #[serde(default = "default_true")]
    pub remove_forms: bool,
}

fn default_true() -> bool {
    true
}

impl From<WasmPreprocessingOptions> for PreprocessingOptionsUpdate {
    fn from(val: WasmPreprocessingOptions) -> Self {
        Self {
            enabled: Some(val.enabled),
            preset: val.preset.map(Into::into),
            remove_navigation: Some(val.remove_navigation),
            remove_forms: Some(val.remove_forms),
        }
    }
}

/// Main conversion options
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct WasmConversionOptions {
    /// Heading style
    pub heading_style: Option<WasmHeadingStyle>,
    /// List indentation type
    pub list_indent_type: Option<WasmListIndentType>,
    /// List indentation width (spaces)
    pub list_indent_width: Option<usize>,
    /// Bullet characters for unordered lists
    pub bullets: Option<String>,
    /// Symbol for strong/emphasis (* or _)
    pub strong_em_symbol: Option<char>,
    /// Escape asterisks in text
    pub escape_asterisks: Option<bool>,
    /// Escape underscores in text
    pub escape_underscores: Option<bool>,
    /// Escape misc markdown characters
    pub escape_misc: Option<bool>,
    /// Escape all ASCII punctuation
    pub escape_ascii: Option<bool>,
    /// Default code language
    pub code_language: Option<String>,
    /// Use autolinks for bare URLs
    pub autolinks: Option<bool>,
    /// Add default title if none exists
    pub default_title: Option<bool>,
    /// Use <br> in tables instead of spaces
    pub br_in_tables: Option<bool>,
    /// Highlight style for `<mark>` elements
    pub highlight_style: Option<WasmHighlightStyle>,
    /// Extract metadata from HTML
    pub extract_metadata: Option<bool>,
    /// Whitespace handling mode
    pub whitespace_mode: Option<WasmWhitespaceMode>,
    /// Strip newlines from HTML before processing
    pub strip_newlines: Option<bool>,
    /// Enable text wrapping
    pub wrap: Option<bool>,
    /// Text wrap width
    pub wrap_width: Option<usize>,
    /// Treat block elements as inline
    pub convert_as_inline: Option<bool>,
    /// Subscript symbol
    pub sub_symbol: Option<String>,
    /// Superscript symbol
    pub sup_symbol: Option<String>,
    /// Newline style
    pub newline_style: Option<WasmNewlineStyle>,
    /// Code block style
    pub code_block_style: Option<WasmCodeBlockStyle>,
    /// Elements where images should remain as markdown
    pub keep_inline_images_in: Option<Vec<String>>,
    /// Skip images during conversion
    pub skip_images: Option<bool>,
    /// Preprocessing options
    pub preprocessing: Option<WasmPreprocessingOptions>,
    /// Source encoding (informational)
    pub encoding: Option<String>,
    /// Enable debug mode with diagnostic warnings
    pub debug: Option<bool>,
    /// List of HTML tags to strip
    pub strip_tags: Option<Vec<String>>,
    /// List of HTML tags to preserve as-is in the output
    pub preserve_tags: Option<Vec<String>>,
    /// Link rendering style
    pub link_style: Option<WasmLinkStyle>,
    /// Output format for conversion
    pub output_format: Option<WasmOutputFormat>,
    /// Include structured document tree in result
    pub include_document_structure: Option<bool>,
    /// Extract inline images from data URIs and SVGs
    pub extract_images: Option<bool>,
    /// Maximum decoded image size in bytes
    pub max_image_size: Option<u64>,
    /// Capture SVG elements as images
    pub capture_svg: Option<bool>,
    /// Infer image dimensions from data
    pub infer_dimensions: Option<bool>,
    /// Maximum DOM tree depth to recurse into
    pub max_depth: Option<usize>,
}

impl From<WasmConversionOptions> for ConversionOptionsUpdate {
    fn from(val: WasmConversionOptions) -> Self {
        Self {
            heading_style: val.heading_style.map(Into::into),
            list_indent_type: val.list_indent_type.map(Into::into),
            list_indent_width: val.list_indent_width,
            bullets: val.bullets,
            strong_em_symbol: val.strong_em_symbol,
            escape_asterisks: val.escape_asterisks,
            escape_underscores: val.escape_underscores,
            escape_misc: val.escape_misc,
            escape_ascii: val.escape_ascii,
            code_language: val.code_language,
            autolinks: val.autolinks,
            default_title: val.default_title,
            br_in_tables: val.br_in_tables,
            highlight_style: val.highlight_style.map(Into::into),
            extract_metadata: val.extract_metadata,
            whitespace_mode: val.whitespace_mode.map(Into::into),
            strip_newlines: val.strip_newlines,
            wrap: val.wrap,
            wrap_width: val.wrap_width,
            convert_as_inline: val.convert_as_inline,
            sub_symbol: val.sub_symbol,
            sup_symbol: val.sup_symbol,
            newline_style: val.newline_style.map(Into::into),
            code_block_style: val.code_block_style.map(Into::into),
            keep_inline_images_in: val.keep_inline_images_in,
            skip_images: val.skip_images,
            preprocessing: val.preprocessing.map(Into::into),
            encoding: val.encoding,
            debug: val.debug,
            strip_tags: val.strip_tags,
            preserve_tags: val.preserve_tags,
            link_style: val.link_style.map(Into::into),
            output_format: val.output_format.map(Into::into),
            include_document_structure: val.include_document_structure,
            extract_images: val.extract_images,
            max_image_size: val.max_image_size,
            capture_svg: val.capture_svg,
            infer_dimensions: val.infer_dimensions,
            max_depth: val.max_depth,
        }
    }
}

impl From<WasmConversionOptions> for html_to_markdown_rs::ConversionOptions {
    fn from(val: WasmConversionOptions) -> Self {
        html_to_markdown_rs::ConversionOptions::from(ConversionOptionsUpdate::from(val))
    }
}
